File: oa2-examples/python/wallet.py
#!/usr/bin/env python3
"""
Resolve an OpenAlias and pay the highest-priority address this wallet supports.

Selects among the recipient's `_openalias-payment` records using the
DNSSEC-authenticated resolution in openalias.py, shows the recipient's display
name (from `_openalias-metadata`), and confirms the address before "sending".

Requires: dnspython with cryptography  ->  pip install "dnspython[dnssec]"

Usage:
    python3 python-wallet.py donate@openalias.org
"""

import sys

import openalias

# What this wallet can pay, as (network, asset) pairs:
#   ("net", None)  -> only the network's native asset
#   ("net", "sym") -> one specific asset on that network
#   ("net", "*")   -> any asset on that network (e.g. a wallet that accepts
#                     the native coin and every token on it)
SUPPORTED = {
    ("btc", None),    # native BTC only
    ("xmr", None),    # native XMR only
    ("eth", "*"),     # any asset on Ethereum (ETH and all tokens)
    ("trx", "usdt"),  # only USDT on Tron
}


def is_supported(record: openalias.PaymentRecord) -> bool:
    """True if this wallet can pay `record` (see SUPPORTED for the conventions)."""
    network = record.network
    if (network, "*") in SUPPORTED:  # the wallet accepts any asset on this network
        return True
    # Collapse a native-asset payment to None so it matches a ("net", None)
    # entry. Nativeness comes from the OA2 network list, so a native asset whose
    # id differs from the network code (e.g. eth on base) is still recognized.
    asset = None if openalias.is_native(record) else record.asset
    return (network, asset) in SUPPORTED


def main(alias: str) -> int:
    try:
        found = openalias.payments(alias)
    except openalias.NotAnAlias as exc:
        print(exc)
        return 1
    except openalias.InsecureLookup as exc:
        print(f"Refusing an unauthenticated result: {exc}")
        return 1
    except openalias.LookupFailed as exc:
        print(f"Lookup failed: {exc}")
        return 1

    if not found:
        print(f"No OpenAlias v2 payment records for {alias}.")
        return 1

    payable = [record for record in found if is_supported(record)]  # already priority-sorted
    if not payable:
        offered = ", ".join(sorted({record.network for record in found}))
        print(f"{alias} accepts {offered}; this wallet supports none of them.")
        return 1

    chosen = payable[0]

    # Metadata is optional and must never block a payment.
    name = None
    try:
        meta = openalias.metadata(alias)
        name = meta.get("name") if meta else None
    except openalias.OpenAliasError:
        pass

    print(f"Paying {name or alias}")
    print(f"  network:      {chosen.network}")
    if chosen.asset:
        print(f"  asset:        {chosen.asset}")
    if chosen.address_type:
        print(f"  address type: {chosen.address_type}")
    print(f"  address:      {chosen.address}")
    if chosen.amount is not None:
        print(f"  amount:       {chosen.amount} {chosen.effective_asset}")
    if chosen.memo:
        print(f"  memo:         {chosen.memo}")

    # Confirm the resolved address with the user before sending.
    if input("\nSend to this address? [y/N] ").strip().lower() == "y":
        print("... (the wallet would build and broadcast the transaction here)")
    else:
        print("Cancelled.")
    return 0


if __name__ == "__main__":
    if len(sys.argv) != 2:
        sys.exit(f"usage: {sys.argv[0]} <alias>   e.g. donate@openalias.org")
    sys.exit(main(sys.argv[1]))

File: oa2-examples/dart/wallet.dart
// Resolve an OpenAlias and pay the highest-priority address this wallet supports.
//
// A Dart port of the Python wallet example. Requires the native helper to be
// built first: `cd native && cargo build --release`.
//
// Usage:
//   dart run wallet.dart donate@openalias.org

import 'dart:io';

import 'openalias.dart' as oa;

// What this wallet can pay, as (network, asset) pairs:
//   (net, null)  -> only the network's native asset
//   (net, 'sym') -> one specific asset on that network
//   (net, '*')   -> any asset on that network
const supported = <(String, String?)>{
  ('btc', null), // native BTC only
  ('xmr', null), // native XMR only
  ('eth', '*'), // any asset on Ethereum (ETH and all tokens)
  ('trx', 'usdt'), // only USDT on Tron
};

bool isSupported(oa.PaymentRecord record) {
  if (supported.contains((record.network, '*'))) return true;
  // Collapse a native-asset payment to null so it matches a (net, null) entry,
  // using the OA2 network list's native-asset mapping.
  final asset = oa.isNative(record) ? null : record.asset;
  return supported.contains((record.network, asset));
}

void main(List<String> args) {
  if (args.length != 1) {
    stderr.writeln('usage: dart run wallet.dart <alias>   e.g. donate@openalias.org');
    exit(1);
  }
  final alias = args[0];

  final List<oa.PaymentRecord> found;
  try {
    found = oa.payments(alias);
  } on oa.NotAnAlias catch (e) {
    print(e);
    exit(1);
  } on oa.InsecureLookup catch (e) {
    print('Refusing an unauthenticated result: $e');
    exit(1);
  } on oa.LookupFailed catch (e) {
    print('Lookup failed: $e');
    exit(1);
  }

  if (found.isEmpty) {
    print('No OpenAlias v2 payment records for $alias.');
    exit(1);
  }

  final payable = found.where(isSupported).toList(); // already priority-sorted
  if (payable.isEmpty) {
    final offered = (found.map((r) => r.network).toSet().toList()..sort()).join(', ');
    print('$alias accepts $offered; this wallet supports none of them.');
    exit(1);
  }

  final chosen = payable.first;

  String? name;
  try {
    name = oa.metadata(alias)?['name'];
  } on oa.OpenAliasException {
    // metadata is optional and must never block a payment
  }

  print('Paying ${name ?? alias}');
  print('  network:      ${chosen.network}');
  if (chosen.asset != null) print('  asset:        ${chosen.asset}');
  if (chosen.addressType != null) print('  address type: ${chosen.addressType}');
  print('  address:      ${chosen.address}');
  if (chosen.amount != null) print('  amount:       ${chosen.amount} ${chosen.effectiveAsset}');
  if (chosen.memo != null) print('  memo:         ${chosen.memo}');

  stdout.write('\nSend to this address? [y/N] ');
  final answer = stdin.readLineSync()?.trim().toLowerCase();
  if (answer == 'y') {
    print('... (the wallet would build and broadcast the transaction here)');
  } else {
    print('Cancelled.');
  }
}
